frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn reveal(
+			origin: OriginFor<T>,
+			name: Vec<u8>,
+			secret: u64,
+			periods: u32,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let commitment_hash = sp_io::hashing::blake2_256(&(&name, secret).encode());
+
+			let commitment =
+				Commitments::<T>::get(commitment_hash).ok_or(Error::<T>::CommitmentNotFound)?;
+
+			let name_hash = sp_io::hashing::blake2_256(&name);
+			let block_number = frame_system::Pallet::<T>::block_number();
+
+			ensure!(
+				block_number > commitment.when.saturating_add(T::MinimumCommitmentPeriod::get()),
+				Error::<T>::TooEarlyToReveal
+			);
+
+			if Self::is_available(name_hash, block_number) {
+				let fee = Self::registration_fee(name.clone(), periods);
+
+				let imbalance = T::Currency::withdraw(
+					&sender,
+					fee,
+					WithdrawReasons::FEE,
+					ExistenceRequirement::KeepAlive,
+				)?;
+
+				T::RegistrationFeeHandler::on_unbalanced(imbalance);
+
+				let deposit = T::NameDeposit::get();
+				Self::do_register(name_hash, commitment.who.clone(), deposit, periods)?;
+			}
+
+			T::Currency::unreserve(&commitment.who, commitment.deposit);
+			Commitments::<T>::remove(commitment_hash);
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			name_hash: NameHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let block_number = frame_system::Pallet::<T>::block_number();
+
+			Registrations::<T>::try_mutate(name_hash, |maybe_registration| {
+				let r = maybe_registration.as_mut().ok_or(Error::<T>::RegistrationNotFound)?;
+				ensure!(r.owner == sender, Error::<T>::NotRegistrationOwner);
+				ensure!(r.expiry > block_number, Error::<T>::RegistrationExpired);
+
+				r.owner = to.clone();
+
+				// TODO: transfer deposit over?
+
+				Self::deposit_event(Event::<T>::Transfer { from: sender, to });
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(0)]
+		pub fn renew(origin: OriginFor<T>, name_hash: NameHash, periods: u32) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			Registrations::<T>::try_mutate(name_hash, |maybe_registration| {
+				let r = maybe_registration.as_mut().ok_or(Error::<T>::RegistrationNotFound)?;
+
+				let fee = Self::length_fee(periods);
+
+				// withdraw fees from account
+				let imbalance = T::Currency::withdraw(
+					&sender,
+					fee,
+					WithdrawReasons::FEE,
+					ExistenceRequirement::KeepAlive,
+				)?;
+
+				let block_number = frame_system::Pallet::<T>::block_number();
+
+				let expiry_new = match r.expiry > block_number {
+					true => r.expiry.saturating_add(Self::length(periods)),
+					false => block_number.saturating_add(Self::length(periods)),
+				};
+
+				r.expiry = expiry_new;
+
+				T::RegistrationFeeHandler::on_unbalanced(imbalance);
+
+				Self::deposit_event(Event::<T>::NameRenewed { name_hash, expires: expiry_new });
```

Comment:
I think events should only be emitted for state changes.  
Could make this function a no-op and return with an `AlreadySet` error if the the new address is the same one as the old.